tests/functional/test_fusion_volume.py

```diff
@@ -0,0 +1,676 @@
+        ),
+        (
+            "tenant",
+            "missing required arguments: tenant",
+        ),
+        (
+            "tenant_space",
+            "missing required arguments: tenant_space",
+        ),
+        (
+            "storage_class",
+            "missing parameter\\(s\\) required by 'placement_group': storage_class",
+        ),
+        (
+            "placement_group",
+            "missing required arguments: placement_group",
+        ),
+        (
+            "size",
+            "missing required arguments: size",
+        ),
+    ],
+)
+def test_module_fails_on_missing_parameters(
+    volumes_api, field, expected_exception_regex, module_args
+):
+    volumes_api = purefusion.VolumesApi()
+    volumes_api.get_volume = MagicMock(side_effect=purefusion.rest.ApiException)
+    del module_args[field]
+    set_module_args(module_args)
+    # run module
+    with pytest.raises(AnsibleFailJson) as ansible_error:
+        fusion_volume.main()
+    assert ansible_error.match(expected_exception_regex)
+
+
+@patch("fusion.VolumesApi")
+@pytest.mark.parametrize(
+    "dict_update,expected_exception_regex",
+    [
+        (
+            {"extra": "value"},
+            "Unsupported parameters for.*module: extra",
+        ),
+        (
+            {"state": "absent"},
+            "Volume must have no host access policies when destroyed",
+        ),
+        (
+            {"eradicate": True},
+            "'eradicate: true' cannot be used together with 'state: present'",
+        ),
+        (
+            {"size": "1K"},
+            "Size is not within the required range",
+        ),
+    ],
+)
+def test_module_fails_on_incorrect_parameters(
+    volumes_api, dict_update, expected_exception_regex, module_args
+):
+    volumes_api = purefusion.VolumesApi()
+    volumes_api.get_volume = MagicMock(side_effect=purefusion.rest.ApiException)
+    module_args.update(dict_update)
+    set_module_args(module_args)
+    # run module
+    with pytest.raises(AnsibleFailJson) as ansible_error:
+        fusion_volume.main()
+    assert ansible_error.match(expected_exception_regex)
+
+
+@patch("fusion.VolumesApi")
+def test_module_not_existent_volume_with_state_absent_not_changed(
+    volumes_api, module_args
+):
+    volumes_api = purefusion.VolumesApi()
+    volumes_api.get_volume = MagicMock(side_effect=purefusion.rest.ApiException)
+    del module_args["host_access_policies"]
+    module_args["state"] = "absent"
+    set_module_args(module_args)
+    # run module
+    with pytest.raises(AnsibleExitJson) as exception:
+        fusion_volume.main()
+    assert exception.value.args[0]["changed"] is False
+
+
+@patch("fusion.OperationsApi")
+@patch("fusion.VolumesApi")
+def test_volume_create_successfully(volumes_api, operations_api, module_args):
+    operations_api = purefusion.OperationsApi()
+    volumes_api = purefusion.VolumesApi()
+    volumes_api.get_volume = MagicMock(side_effect=purefusion.rest.ApiException)
```

Comment:
Please make sure that `get_volume` method was actually called. The same for the rest of the tests, whenever something is mocked, we can take advantage of checking how many times it was called or if it was called at all. If `get_volume` wasn't called, then this test should fail as the module didn't make sure whether the volume have existed before.

The same for `get_operation` method.